stdlib/universe/narrow_state_tracking.go

```diff
@@ -0,0 +1,237 @@
+	"github.com/influxdata/flux/values"
+)
+
+func NewNarrowStateTrackingTransformation(ctx context.Context, spec *StateTrackingProcedureSpec, id execute.DatasetID, mem memory.Allocator) (execute.Transformation, execute.Dataset, error) {
+	fn := execute.NewRowPredicateFn(spec.Fn.Fn, compiler.ToScope(spec.Fn.Scope))
+	a := &narrowStateTrackingTransformationAdapter{
+		ctx:      ctx,
+		fn:       fn,
+		timeCol:  spec.TimeCol,
+		countCol: spec.CountColumn,
+		durCol:   spec.DurationColumn,
+		unit:     int64(spec.DurationUnit.Duration()),
+	}
+	nt, d, err := execute.NewNarrowStateTransformation(id, a, mem)
+	if err != nil {
+		return nil, nil, err
+	}
+	return nt, d, nil
+}
+
+type narrowStateTrackingTransformationAdapter struct {
+	ctx context.Context
+	fn  *execute.RowPredicateFn
+
+	timeCol,
+	countCol,
+	durCol string
+
+	unit int64
+}
+
+type trackedState struct {
+	start,
+	prevTime values.Time
+
+	countInState,
+	durationInState bool
+
+	count,
+	duration int64
+}
+
+func (a *narrowStateTrackingTransformationAdapter) Process(chunk table.Chunk, state interface{}, d *execute.TransportDataset, mem memory.Allocator) (interface{}, bool, error) {
+	mod := false // Track whether or not the state has been modified
+	if chunk.Len() == 0 {
+		chunk.Retain()
+		d.Process(chunk)
+		return state, mod, nil
+	}
+
+	if state == nil {
+		state = trackedState{
+			count:    -1,
+			duration: -1,
+		}
+		mod = true
+	}
+	s := state.(trackedState)
+
+	var (
+		times     array.Int
+		fn        execute.RowPredicatePreparedFn
+		counts    array.IntBuilder
+		durations array.IntBuilder
+	)
+	if err := a.scanChunk(chunk, mem, &times, &fn, &counts, &durations); err != nil {
+		chunk.Retain()
+		d.Process(chunk)
+		return s, mod, err
+	}
+
+	buf := chunk.Buffer()
+	for i := 0; i < chunk.Len(); i++ {
+		match, err := fn.EvalRow(a.ctx, i, &buf)
+		if err != nil {
+			chunk.Retain()
+			d.Process(chunk)
+			return s, mod, err
+		}
+		ts := values.Time(times.Value(i))
+		if mod, err := a.updateState(&s, match, ts, &counts, &durations); err != nil {
+			chunk.Retain()
+			d.Process(chunk)
+			return s, mod, err
+		}
+	}
+
+	chunk.Retain()
+	c := a.appendStateColumns(chunk, &counts, &durations)
+	if err := d.Process(c); err != nil {
+		return s, mod, err
+	}
+	return s, mod, nil
+}
+
+func (a *narrowStateTrackingTransformationAdapter) Close() error { return nil }
+
+func (a *narrowStateTrackingTransformationAdapter) scanChunk(
+	chunk table.Chunk,
+	mem memory.Allocator,
+	times *array.Int,
+	fn *execute.RowPredicatePreparedFn,
+	counts *array.IntBuilder,
+	durations *array.IntBuilder,
+) error {
+	timeIdx := chunk.Index(a.timeCol)
+	if timeIdx < 0 {
+		return errors.Newf(codes.FailedPrecondition, "column %q does not exist", a.timeCol)
+	}
+
+	buf := chunk.Buffer()
+	*times = *buf.Times(timeIdx)
+
+	if a.durCol != "" {
+		if times.NullN() > 0 {
+			return errors.New(codes.FailedPrecondition, "got a null timestamp")
+		}
+	}
+
+	*counts = *array.NewIntBuilder(mem)
```

Comment:
This is a C-ism that we should likely avoid. It doesn't seem like these are used in here, but if it logically makes sense to do this here, remember that it's ok for this method to call another method.

When we want to break up a process method into multiple methods logically, we can do that through a function that calls another function. If it makes sense to initialize the builders here, we can use Go's support for multiple return values to return the state we want to pass to another function. Both of these are valid. So, for example, this function might return the two builders, the prepared function, and an error if an error happens.